### PR TITLE
DTSERWFOUR-2 Transaction Screen - Display Prepaid Card Transactions

### PR DESCRIPTION
**Pull request by rajkumar-balakrishnasamy**

### CODE REVIEW

File: receiptrepository/src/main/java/com/hyperwallet/android/ui/receipt/repository/PrepaidCardRepository.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import com.hyperwallet.android.model.Errors;
+import com.hyperwallet.android.model.transfermethod.PrepaidCard;
+
+import java.util.List;
+
+/**
+ * Prepaid card Repository Contract
+ */
+public interface PrepaidCardRepository {
```
Comment (vwagh-hw) [inline receiptrepository/src/main/java/com/hyperwallet/android/ui/receipt/repository/PrepaidCardRepository.java:30]: wrong place, this should be in transfer method repository module as I mentioned before. If you want you can actually point to @shanavas123's branch and you will have this repository

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/HyperwalletReceiptUi.java [hunk 2]
```diff
@@ -52,11 +52,11 @@ public static synchronized HyperwalletReceiptUi getInstance(
      * @param lockScreenToPortrait if set <code>true</code> screen will be locked to Portrait mode;
      *                             otherwise <code>false</code> screen will follow whatever the
      *                             device orientation is directed.
-     * @return an Intent with the data necessary to launch the {@link ListUserReceiptActivity}
+     * @return an Intent with the data necessary to launch the {@link TabbedListReceiptActivity}
      */
     public Intent getIntentListUserReceiptActivity(@NonNull final Context context, final boolean lockScreenToPortrait) {
-        Intent intent = new Intent(context, ListUserReceiptActivity.class);
-        intent.putExtra(ListUserReceiptActivity.EXTRA_LOCK_SCREEN_ORIENTATION_TO_PORTRAIT, lockScreenToPortrait);
+        Intent intent = new Intent(context, TabbedListReceiptActivity.class);
```
Comment (vwagh-hw) [inline receiptui/src/main/java/com/hyperwallet/android/ui/receipt/HyperwalletReceiptUi.java:58]: Why are you updating the existing method for user receipts? Please refer to the design diagram, we are creating a new method for tabbed view..

Comment (rajkumar-balakrishnasamy) [reply receiptui/src/main/java/com/hyperwallet/android/ui/receipt/HyperwalletReceiptUi.java:58]: Sorry. Updated with new method.

File: transfermethodrepository/src/main/java/com/hyperwallet/android/ui/transfermethod/repository/PrepaidCardRepository.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+import com.hyperwallet.android.model.Errors;
+import com.hyperwallet.android.model.transfermethod.PrepaidCard;
+
+import java.util.List;
+
+/**
+ * Prepaid card Repository Contract
+ */
+public interface PrepaidCardRepository {
```
Comment (vwagh-hw) [inline transfermethodrepository/src/main/java/com/hyperwallet/android/ui/transfermethod/repository/PrepaidCardRepository.java:30]: Note to reviewers and QA, After create transfer is merged this and related code can be removed (will conflict) 
its a copy/paste for this branch to work.

File: receiptui/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -165,4 +165,9 @@
     <string name="fispc">Purchase</string>
     <string name="visa">Visa</string>
     <string name="mastercard">Mastercard</string>
+
+    <!-- Prepaid Cards Receipt   -->
+    <string name="mobileAvailableBalances">Available funds</string>
+    <string name="prepaidCardTabLabel">%1$s%2$s</string>
```
Comment (vwagh-hw) [inline receiptui/src/main/res/values/strings.xml:171]: there is no string here to translate, why is this here

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/TabbedListReceiptsFragment.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+            @Override
+            public int compare(PrepaidCard firstPrepaid, PrepaidCard secondPrepaid) {
+                if (firstPrepaid == null || secondPrepaid == null
+                        || firstPrepaid.getPrimaryCardToken() == null || secondPrepaid.getPrimaryCardToken() == null) {
+                    return 0;
+                }
+                return firstPrepaid.getPrimaryCardToken().compareTo(secondPrepaid.getPrimaryCardToken());
+            }
+        });
+    }
```
Comment (vwagh-hw) [inline receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/TabbedListReceiptsFragment.java:162]: method name doesn't match its behavior, it should be `sortPrepaidCards`
Simplify the logic here, refer to how its done in App

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/TabbedListReceiptsFragment.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        mTabbedListReceiptsViewModel.mPrepaidCards.observe(this, new Observer<List<PrepaidCard>>() {
+            @Override
+            public void onChanged(List<PrepaidCard> prepaidCards) {
+                if (!prepaidCards.isEmpty()) {
+                    boolean isWalletModel =
+                            mTabbedListReceiptsViewModel.mProgramModel.name() == ProgramModel.WALLET_MODEL.name();
+                    boolean isCardModel =
+                            mTabbedListReceiptsViewModel.mProgramModel.name() == ProgramModel.PAY2CARD_MODEL.name()
+                                    || mTabbedListReceiptsViewModel.mProgramModel.name()
+                                    == ProgramModel.CARD_ONLY_MODEL.name();
```
Comment (vwagh-hw) [inline receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/TabbedListReceiptsFragment.java:122]: this logic can be moved to ProgramModel, refer to same comment in create transfer PR

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/TabbedListReceiptsFragment.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    @Override
+    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+        mViewPager = view.findViewById(R.id.receipts_pager);
+        mTabLayout = view.findViewById(R.id.tab_layout);
+        mHeader = view.findViewById(R.id.transactions_header);
+    }
+
+    void retry() {
+        // mReceiptViewModel.retry();
```
Comment (vwagh-hw) [inline receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/TabbedListReceiptsFragment.java:89]: Please make sure you don't leave any commented code

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/ListReceiptsFragment.java [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+    @Override
+    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
+        if (savedInstanceState != null) {
+            mShouldShowNoTransactionPlaceholder = savedInstanceState.getBoolean(SHOULD_SHOW_NO_TRANSACTION_PLACEHOLDER);
+        }
+        super.onViewStateRestored(savedInstanceState);
+        if (savedInstanceState != null) {
+            mPrepaidCardToken = savedInstanceState.getString(ARGUMENT_PREPAID_CARD_TOKEN);
+        } else {
+            //   mPrepaidCardToken = getArguments().getString(ARGUMENT_PREPAID_CARD_TOKEN);
```
Comment (peter-joseph) [inline receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/ListReceiptsFragment.java:138]: remove this case if not used

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/viewmodel/ProgramModel.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package com.hyperwallet.android.ui.receipt.viewmodel;
+
+public enum ProgramModel {
```
Comment (peter-joseph) [inline receiptui/src/main/java/com/hyperwallet/android/ui/receipt/viewmodel/ProgramModel.java:3]: add this in `common` module

File: transfermethodrepository/src/main/java/com/hyperwallet/android/ui/transfermethod/repository/PrepaidCardRepository.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+     * Load prepaid cards information
+     * <p>
+     * When prepaid cards is properly loaded
+     * {@link LoadPrepaidCardsCallback#onPrepaidCardLoaded(List<PrepaidCard>))}
+     * is invoked otherwise {@link LoadPrepaidCardsCallback#onError(Errors)}
+     * is called to further log or show error information
+     */
+    interface LoadPrepaidCardsCallback {
+
+        void onPrepaidCardLoaded(@NonNull final List<PrepaidCard> prepaidCardList);
```
Comment (peter-joseph) [inline transfermethodrepository/src/main/java/com/hyperwallet/android/ui/transfermethod/repository/PrepaidCardRepository.java:49]: from `onPrepaidCardLoaded` to `onPrepaidCardsLoaded`

File: transfermethodrepository/src/main/java/com/hyperwallet/android/ui/transfermethod/repository/PrepaidCardRepositoryImpl.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public void loadPrepaidCards(@NonNull final LoadPrepaidCardsCallback callback) {
+
+        PrepaidCardQueryParam queryParam = new PrepaidCardQueryParam.Builder().status(ACTIVATED).build();
+
+        getHyperwallet().listPrepaidCards(queryParam, new HyperwalletListener<PageList<PrepaidCard>>() {
+            @Override
+            public void onSuccess(@Nullable PageList<PrepaidCard> result) {
+                callback.onPrepaidCardLoaded(result.getDataList());
```
Comment (peter-joseph) [inline transfermethodrepository/src/main/java/com/hyperwallet/android/ui/transfermethod/repository/PrepaidCardRepositoryImpl.java:53]: from `onPrepaidCardLoaded` to `onPrepaidCardsLoaded`

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/ListReceiptsFragment.java [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+                    public void onChanged(int position, int count) {
+                    }
+
+                    @Override
+                    public void onInserted(int position, int count) {
+                        if (receipts.size() > 0) {
+                            mShouldShowNoTransactionPlaceholder = Boolean.FALSE;
+                        } else {
+                            mShouldShowNoTransactionPlaceholder = Boolean.TRUE;
+                        }
```
Comment (vwagh-hw) [inline receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/ListReceiptsFragment.java:158]: you can use ternary operator for such simple if/else

`mShouldShowNoTransactionPlaceholder = receipts.size() > 0 ? Boolean.FALSE : Boolean.TRUE;`

**Comment by vwagh-hw** [comment]
@rajkumar-balakrishnasamy receipts should show most recent first, currently they are showing incorrect order.

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/viewmodel/TabbedListReceiptsViewModel.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+            }
+        });
+    }
+
+    private ProgramModel getProgramModel() {
+        Hyperwallet.getDefault().getConfiguration(new HyperwalletListener<Configuration>() {
+            @Override
+            public void onSuccess(@Nullable Configuration result) {
+                if (result != null) {
+                    mProgramModel = ProgramModel.valueOf(result.getProgramModel());
```
Comment (shanavas123) [inline receiptui/src/main/java/com/hyperwallet/android/ui/receipt/viewmodel/TabbedListReceiptsViewModel.java:84]: check if result.getProgranModel() is not as empty string, if its empty its caues crash.

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/viewmodel/ProgramModel.java (hunk 9 above)
Comment (rajkumar-balakrishnasamy) [reply receiptui/src/main/java/com/hyperwallet/android/ui/receipt/viewmodel/ProgramModel.java:3]: @peter-joseph : Moved ProgramModel to the 'view' package in common-ui.

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/TabbedListReceiptsFragment.java (hunk 5 above)
Comment (rajkumar-balakrishnasamy) [reply receiptui/src/main/java/com/hyperwallet/android/ui/receipt/view/TabbedListReceiptsFragment.java]: Can't skip the null checks and to simply this implementation Java8 is required which is not supported below Android 7.0. Should introduce comparator to new class?

File: receiptui/src/main/res/values/strings.xml (hunk 4 above)
Comment (rajkumar-balakrishnasamy) [reply receiptui/src/main/res/values/strings.xml:171]: Fixed

**Review by shanavas123 — DISMISSED** [review]

**Review by pdamodarannair — APPROVED** [review]

**Review by vwagh-hw — COMMENTED** [review]
This PR needs to fix compilation issues, code coverage fixes to make TC happy (and fix sorting if not already done)
Once TC is green, @pdamodarannair please review/approve this PR and get it merged

**Review by shanavas123 — DISMISSED** [review]

File: receiptui/src/test/java/com/hyperwallet/android/ui/receipt/viewmodel/TabbedListReceiptsViewModelTest.java [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+            @Override
+            public Object answer(InvocationOnMock invocation) {
+                PrepaidCardRepository.LoadPrepaidCardsCallback callback = invocation.getArgument(0);
+                callback.onPrepaidCardListLoaded(mPrepaidCardList);
+                return callback;
+            }
+        }).when(mPrepaidCardRepository).loadPrepaidCards(any(
+                PrepaidCardRepository.LoadPrepaidCardsCallback.class));
+
+
```
Comment (vwagh-hw) [inline receiptui/src/test/java/com/hyperwallet/android/ui/receipt/viewmodel/TabbedListReceiptsViewModelTest.java:221]: 2 empty lines, delete 1

File: receiptui/src/test/java/com/hyperwallet/android/ui/receipt/viewmodel/ListUserReceiptsViewModelTest.java [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Test
+    public void testGetReceiptList_returnsLiveData() {
+        assertThat(mListUserReceiptsViewModel.receipts(), is(notNullValue()));
+    }
+
+    @Test
+    public void testGetDetailNavigation_returnsLiveData() {
+        MatcherAssert.assertThat(mListUserReceiptsViewModel.getDetailNavigation(),
+                CoreMatchers.is(CoreMatchers.notNullValue()));
```
Comment (vwagh-hw) [inline receiptui/src/test/java/com/hyperwallet/android/ui/receipt/viewmodel/ListUserReceiptsViewModelTest.java:53]: use static imports, Don't use `MatcherAssert.assertThat` or `CoreMatchers.is` or `CoreMatchers.notNullValue()`
it should be simple as `assertThat(` or `is(` or `notNullValue(`
Revisit all your test code and fix wherever applicable

File: receiptui/src/test/java/com/hyperwallet/android/ui/receipt/viewmodel/ListReceiptsViewModelTest.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Test
+    public void testGetReceiptList_returnsLiveData() {
+        assertThat(mListReceiptsViewModel.receipts(), is(notNullValue()));
+    }
+
+    @Test
+    public void testGetDetailNavigation_returnsLiveData() {
+        MatcherAssert.assertThat(mListReceiptsViewModel.getDetailNavigation(), CoreMatchers.is(CoreMatchers.notNullValue()));
```
Comment (vwagh-hw) [inline receiptui/src/test/java/com/hyperwallet/android/ui/receipt/viewmodel/ListReceiptsViewModelTest.java:51]: use static imports plz

File: receiptui/src/test/java/com/hyperwallet/android/ui/receipt/viewmodel/ListPrepaidCardReceiptsViewModelTest.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Test
+    public void testGetReceiptList_returnsLiveData() {
+        assertThat(mListPrepaidCardReceiptsViewModel.receipts(), is(notNullValue()));
+    }
+
+    @Test
+    public void testGetDetailNavigation_returnsLiveData() {
+        MatcherAssert.assertThat(mListPrepaidCardReceiptsViewModel.getDetailNavigation(),
+                CoreMatchers.is(CoreMatchers.notNullValue()));
```
Comment (vwagh-hw) [inline receiptui/src/test/java/com/hyperwallet/android/ui/receipt/viewmodel/ListPrepaidCardReceiptsViewModelTest.java:53]: static imports plz

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/viewmodel/ListPrepaidCardReceiptsViewModel.java [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+        mPrepaidCardReceiptRepository.retryLoadReceipt();
+    }
+
+    @Override
+    public LiveData<Event<Receipt>> getDetailNavigation() {
+        return mDetailNavigation;
+    }
+
+    @Override
+    public void setDetailNavigation(@NonNull Receipt receipt) {
```
Comment (vwagh-hw) [inline receiptui/src/main/java/com/hyperwallet/android/ui/receipt/viewmodel/ListPrepaidCardReceiptsViewModel.java:75]: `receipt` should be final

File: receiptui/src/main/java/com/hyperwallet/android/ui/receipt/viewmodel/ListUserReceiptsViewModel.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+        mUserReceiptRepository.retryLoadReceipt();
+    }
+
+    @Override
+    public LiveData<Event<Receipt>> getDetailNavigation() {
+        return mDetailNavigation;
+    }
+
+    @Override
+    public void setDetailNavigation(@NonNull Receipt receipt) {
```
Comment (vwagh-hw) [inline receiptui/src/main/java/com/hyperwallet/android/ui/receipt/viewmodel/ListUserReceiptsViewModel.java:77]: `receipt` should be final

**Review by malaw-hw — DISMISSED** [review]

**Review by shanavas123 — APPROVED** [review]

**Review by shanavas123 — APPROVED** [review]

**Review by malaw-hw — APPROVED** [review]
